test/e2e_node/gcloud/gcloud.go

```diff
@@ -0,0 +1,174 @@
+	args = append(args, aa...)
+	glog.Infof("Command %s", strings.Join(args, " "))
+	return exec.Command("gcloud", args...).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyToHost(from string, to string) ([]byte, error) {
+	rto := fmt.Sprintf("%s:%s", gc.host, to)
+	glog.Infof("Command gcloud compute copy-files --zone %s %s %s", gc.zone, from, rto)
+	return exec.Command("gcloud", "compute", "copy-files", "--zone", gc.zone, from, rto).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyAndRunAsync(sudo bool, remotePort string, bin string, args ...string) *CmdHandle {
+	h := &CmdHandle{}
+	h.Output = make(chan RunResult)
+
+	rand.Seed(time.Now().UnixNano())
+
+	// Define where we will copy the temp binary
+	tDir := fmt.Sprintf("/tmp/gcloud-e2e-%d", rand.Int31())
+	_, f := filepath.Split(bin)
+	cmd := filepath.Join(tDir, f)
+	h.LPort = getLocalPort()
+
+	// Do the setup
+	go func() {
+		// Create the tmp directory
+		out, err := gc.Command("mkdir", "-p", tDir)
+		if err != nil {
+			glog.Errorf("mkdir failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("mkdir -p %s", tDir)}
+			return
+		}
+
+		// Copy the binary
+		out, err = gc.CopyToHost(bin, tDir)
+		if err != nil {
+			glog.Errorf("copy-files failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("copy-files %s %s", bin, tDir)}
+			return
+		}
+
+		// Start the process
+		out, err = gc.TunnelCommand(sudo, h.LPort, remotePort, cmd, args...)
+		if err != nil {
+			glog.Errorf("command failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("%s %s", cmd, strings.Join(args, " "))}
+			return
+		}
+	}()
+
+	h.TearDown = func() {
+		out, err := gc.Command("sudo", "pkill", f)
+		if err != nil {
+			h.Output <- RunResult{out, err, fmt.Sprintf("pkill %s", cmd)}
+			return
+		}
+		out, err = gc.Command("rm", "-rf", tDir)
+		if err != nil {
+			h.Output <- RunResult{out, err, fmt.Sprintf("rm -rf %s", tDir)}
+			return
+		}
+	}
+	return h
+}
+
+func (gc *gCloudClientImpl) CopyAndWaitTillHealthy(
+	sudo bool,
+	remotePort string, timeout time.Duration, healthUrl string, bin string, args ...string) (*CmdHandle, error) {
+	h := gc.CopyAndRunAsync(sudo, remotePort, bin, args...)
+	eTime := time.Now().Add(timeout)
+	done := false
+	for eTime.After(time.Now()) && !done {
+		select {
+		case r := <-h.Output:
+			glog.Infof("Error running %s Output:\n%s Error:\n%v", r.cmd, r.out, r.err)
+			return h, r.err
+		case <-time.After(2 * time.Second):
+			resp, err := http.Get(fmt.Sprintf("http://localhost:%s/%s", h.LPort, healthUrl))
+			if err == nil && resp.StatusCode == http.StatusOK {
+				done = true
+				break
+			}
+		}
+	}
+	if !done {
+		return h, errors.New(fmt.Sprint("Timeout waiting for service to be healthy at http://localhost:%s/%s", h.LPort, healthUrl))
+	}
+	glog.Info("Healthz Success")
+	return h, nil
+}
+
+// GetLocalPort returns a free local port that can be used for ssh tunneling
+func getLocalPort() string {
+	l, _ := net.Listen("tcp", ":0")
```

Comment:
yeah there is a race condition here if something else allocates the port before we get to use it in ssh.  The goal is to give a free port to "gcloud ssh" to setup a tunnel so that we don't need to worry about firewall rules on the host running the kubelet and so requests come from "localhost".  I wasn't able to find a solution for reserving a free port so it can be used by another process.